src/poetry/console/commands/plugin/add.py

```diff
@@ -74,57 +71,53 @@ def handle(self) -> int:
-            pyproject = tomlkit.loads(
-                env_dir.joinpath("pyproject.toml").read_text(encoding="utf-8")
-            )
-            poetry_content = pyproject["tool"]["poetry"]
-            existing_packages = self.get_existing_packages_from_input(
-                plugins, poetry_content, "dependencies"
+        existing_plugins = {}
+        if env_dir.joinpath("plugins.toml").exists():
+            existing_plugins = tomlkit.loads(
+                env_dir.joinpath("plugins.toml").read_text(encoding="utf-8")
             )
 
-            if existing_packages:
-                self.notify_about_existing_packages(existing_packages)
+        root_package = self.create_env_package(system_env, existing_plugins)
 
-            plugins = [plugin for plugin in plugins if plugin not in existing_packages]
+        installed_plugins = {
+            canonicalize_name(ep.distro.name) for ep in self.get_plugin_entry_points()
+        }
 
-        if not plugins:
-            return 0
+        # We check for the plugins existence first.
+        plugins = []
+        skipped_plugins = []
+        parsed_plugins = self._parse_requirements(requested_plugins)
+        for i, plugin in enumerate(parsed_plugins):
+            plugin_name = canonicalize_name(plugin.pop("name"))
+            if plugin_name in installed_plugins:
+                if plugin_name not in existing_plugins:
+                    existing_plugins[plugin_name] = plugin
+
+                if not plugin:
+                    skipped_plugins.append(plugin_name)
 
-        plugins = self._determine_requirements(plugins)
+                    continue
 
-        # We retrieve the packages installed in the system environment.
-        # We assume that this environment will be a self contained virtual environment
-        # built by the official installer or by pipx.
-        # If not, it might lead to side effects since other installed packages
-        # might not be required by Poetry but still taken into account when resolving dependencies.
-        installed_repository = InstalledRepository.load(
-            system_env, with_dependencies=True
-        )
+            plugins.append(canonicalize_name(requested_plugins[i]))
+
+        if skipped_plugins:
+            self.line(
+                "The following plugins are already present and will be skipped:\n"
+            )
+            for name in sorted(skipped_plugins):
+                self.line(f"  • <c1>{name}</c1>")
 
-        root_package = None
-        for package in installed_repository.packages:
-            if package.name == "poetry":
-                root_package = ProjectPackage(package.name, package.version)
-                for dependency in package.requires:
-                    root_package.add_dependency(dependency)
+            self.line(
+                "\nIf you want to upgrade it to the latest compatible version, "
+                "you can use `poetry plugin add plugin@latest.\n"
```

Comment:
I want to distinguish the `latest` special notation from true versions, hence the `@` notation to treat it differently.

Reply:
Hmm, fair enough. It's mildly surprising, but not totally out of left field since node uses this syntax. I think we should call it out as poetry-specific loudly in the docs then.